### PR TITLE
Set map tool

### PR DESCRIPTION
**Pull request by YoannQDQ**
## Description

When setting a map tool, QGIS does not check if it is already the active tool, and will always deactivate then reactivate the tool. This may cause issues (e.g glitchy MeasureDialog on Ubuntu when user clicks on the measure tool button while it is already active)

This PR modify the `QgsMapCanvas::setMapTool` method, to check whether the new tool is actually the old one. It it is, rather than deactivating and activating the tool, we call `QgsMapTool::reactivate()` and return. `QgsMapTool::reactivate()` default implementation does nothing. The glitchy MeasureDialog on Ubuntu is also adressed in this PR.

### CODE REVIEW

File: src/app/qgsmeasuretool.cpp [hunk 1]
```diff
@@ -111,6 +111,16 @@ void QgsMeasureTool::deactivate()
   QgsMapTool::deactivate();
 }
 
+void QgsMeasureTool::reactivate()
+{
+  // User clicked on the measure action while it was already active
+  // Only ensure that the dialog and rubberbands visible
+  mDialog->show();
+  mRubberBand->show();
+  mRubberBandPoints->show();
```
[the remark this answers is not in the document]
Comment (YoannQDQ) [reply src/app/qgsmeasuretool.cpp]: You're right, they wouldn't.

File: src/gui/qgsmaptool.h [hunk 2]
```diff
@@ -192,6 +192,12 @@ class GUI_EXPORT QgsMapTool : public QObject
     //! called when map tool is being deactivated
     virtual void deactivate();
 
+    /**
+     * called when map tool is being activated while it is already active
```
Comment (nyalldawson) [inline src/gui/qgsmaptool.h:196]:
```suggestion
     * Called when the map tool is being activated while it is already active.
     *
     * The default implementation does nothing.
```

**Review by nyalldawson — COMMENTED** [review]
Good move! Would you mind adding some tests covering this too?

File: src/gui/qgsmaptool.h [hunk 3]
```diff
@@ -192,6 +192,14 @@ class GUI_EXPORT QgsMapTool : public QObject
     //! called when map tool is being deactivated
     virtual void deactivate();
 
+    /**
+     * Called when the map tool is being activated while it is already active.
+     *
+     * The default implementation does nothing.
```
Comment (nyalldawson) [inline src/gui/qgsmaptool.h:198]:
```suggestion
     * The default implementation emits the reactivated () signal.
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Just needs a sipify run...

**Pull request merged**